### PR TITLE
ipv6 lookback issue

### PR DESCRIPTION
**Pull request by Grubba27**
### Pull Request checklist ###

- [x] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [ x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes a changelog entry in [CHANGES_UNRELEASED.md](../CHANGES_UNRELEASED.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [x ] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

This pull request refers to [this issue](https://github.com/mozilla/application-services/issues/4775)
This pull request is just a refactoring of the loopback, I think it does not need an entry in Changes_unreleased.md but if this is the case i would write there too

### CODE REVIEW

File: components/viaduct/src/backend.rs [hunk 1]
```diff
@@ -47,6 +47,8 @@ pub fn validate_request(request: &crate::Request) -> Result<(), crate::Error> {
     if request.url.scheme() != "https"
         && request.url.host_str() != Some("localhost")
         && request.url.host_str() != Some("127.0.0.1")
+        && request.url.host_str() != Some("[::1]")
```
Comment (mhammond) [inline components/viaduct/src/backend.rs:50]: In the issue, I think Martin was suggesting that the `host_str` checks can be replaced with something like:

```rust
match request.url.host() {
    Some(url::Host::Domain(d)) => d != "localhost",
    Some(url::Host::Ipv4(addr)) => !addr.is_loopback(),
    Some(url::Host::Ipv6(addr)) => !addr.is_loopback(),
    None => true,
}
```

(with a `use url` at the top of the file) which is cleaner and handles every possible way to spell an ipv4 and ipv6 loopback addresses)

**Review by mhammond — DISMISSED** [review]
Thanks heaps for digging in here!

File: components/viaduct/src/backend.rs (hunk 1 above)
Comment (Grubba27) [reply components/viaduct/src/backend.rs:50]: i got it but when i do this changes this validation start to fail maybe i did something wrong?
```rust
pub fn validate_request(request: &crate::Request) -> Result<(), crate::Error> {
    match request.url.host() {
        Some(url::Host::Domain(d)) => d != "localhost",
        Some(url::Host::Ipv4(addr)) => !addr.is_loopback(),
        Some(url::Host::Ipv6(addr)) => !addr.is_loopback(),
        None => return Err(crate::Error::NonTlsUrl),
    };
    Ok(())
}
// the only test that fails


        let _http_request = crate::Request::new(
            crate::Method::Get,
            url::Url::parse("http://www.example.com").unwrap(),
        );
        assert!(validate_request(&_http_request).is_err());
```

Comment (mhammond) [reply components/viaduct/src/backend.rs]: You still need the check for `request.url.scheme() != "https"` - it's only the `host_str` checks that can be replaced with that block. That's why my block has `None => true` - ie, something like:

```rust
    if request.url.scheme() != "https"
        && match request.url.host() {
            Some(url::Host::Domain(d)) => d != "localhost",
            Some(url::Host::Ipv4(addr)) => !addr.is_loopback(),
            Some(url::Host::Ipv6(addr)) => !addr.is_loopback(),
            None => true,
        }
    {
        return Err(crate::Error::NonTlsUrl);
    }
```

Comment (Grubba27) [reply components/viaduct/src/backend.rs]: OMG, I've never tought using match this way. This really simple thing that you have shown me made me love even more this language

**Comment by Grubba27** [comment]
I'think that all is ok and I'm realy glad that I learned a new thing

**Review by mhammond — APPROVED** [review]
Thanks!

**Pull request merged**